src/aro/Preprocessor.zig

```diff
@@ -409,19 +409,53 @@ pub fn expansionSlice(pp: *Preprocessor, tok: Tree.TokenIndex) []Source.Location
     return locs[0..i];
 }
 
+const TokenCount = struct {
+    tokens_len: usize,
+    expansion_entries_len: usize,
+};
+
+pub fn getTokenCount(pp: *const Preprocessor) TokenCount {
+    return .{
+        .tokens_len = pp.tokens.len,
+        .expansion_entries_len = pp.expansion_entries.len,
+    };
+}
+
+/// `count` must be less than or equal to the current token count; i.e. this function cannot be used to grow the list
+pub fn setTokenCount(pp: *Preprocessor, count: TokenCount) void {
+    assert(pp.tokens.len >= count.tokens_len);
+    for (pp.expansion_entries.items(.locs)[count.expansion_entries_len..]) |locs| TokenWithExpansionLocs.free(locs, pp.comp.gpa);
+    pp.tokens.len = count.tokens_len;
+    pp.expansion_entries.len = count.expansion_entries_len;
+}
+
 /// Preprocess a compilation unit of sources into a parsable list of tokens.
-pub fn preprocessSources(pp: *Preprocessor, sources: []const Source, implicit_includes: []const Source) Error!void {
+pub fn preprocessSources(pp: *Preprocessor, sources: []const Source, imacros: []const Source, implicit_includes: []const Source) Error!void {
     assert(sources.len > 1);
     const first = sources[0];
+
     try pp.addIncludeStart(first);
+    var prev_source: ?Source = null;
     for (sources[1..]) |header| {
         try pp.addIncludeStart(header);
         _ = try pp.preprocess(header);
+        if (prev_source) |prev| try pp.addIncludeResume(prev.id, 0, 1);
+        prev_source = header;
     }
     pp.include_depth = 1;
+    for (imacros) |imacro| {
+        try pp.addIncludeStart(imacro);
+        const token_count = pp.getTokenCount();
+        _ = try pp.preprocess(imacro);
+        assert(pp.include_depth == 1);
+        pp.setTokenCount(token_count);
+        try pp.addIncludeResume(sources[1].id, 0, 1);
```

Comment:
These should probably use `prev_source` or `first` instead of asserting that there are at least two sources.